File: elixir_solutions/lib/0055.ex
defmodule Solution0055 do
  @spec can_jump(nums :: [integer]) :: boolean
  def can_jump(nums) do
    can_jump(nums, 0, Enum.at(nums, 0))
  end

  def can_jump(nums, current, reachable) when current >= reachable do
    #IO.inspect({current, reachable, false})
    current >= length(nums) - 1
  end
  def can_jump(nums, current, reachable) do
    #IO.inspect({current, reachable})
    l = length(nums) - 1
    if reachable >= l do
      true
    else
      next = current + 1
      next_reachable =
        [Enum.at(nums, next) + next, reachable]
        |> Enum.max
      can_jump(nums, next, next_reachable)
    end
  end
end

File: javascript/0055.js
var canJump = function(nums) {
    let max = 0
    let pos = 0
    while (pos <= max) {
        max = Math.max(pos +nums[pos], max)
        if (max >= nums.length - 1) {
            return true
        }
        pos = pos + 1
    }
    return false
};

nums = [2,3,1,1,4]
console.log(canJump(nums))
nums = [3,2,1,0,4]
console.log(canJump(nums))